### PR TITLE
Add Tokio Mutexes

### PR DESCRIPTION
**Pull request by mihirsamdarshi**
Hey,

Great project! I began adding Tokio mutexes. 

This PR is definitely not ready, so I wanted to see if you had any pointers or tips for me in regards to the following:

One issue I was running into when integrating it into a project that I have is that it reaches the `unreachable!` portion of the code in [`mark_released`](https://github.com/bertptrs/tracing-mutex/blob/5232bac5828d17a6d2e52d4e5091e0877efbd1c7/src/lib.rs#L176). I suspect this is related to your comment in #4, where Mutexes are not able to be tracked properly. I was wondering if you had any tips on how I could solve this?

 I have also not added any documentation or sufficient tests

### CODE REVIEW

**Comment by bertptrs** [comment]
Hey, first of all thank you for your interest in the project!

> One issue I was running into when integrating it into a project that I have is that it reaches the `unreachable!` portion of the code in [`mark_released`](https://github.com/bertptrs/tracing-mutex/blob/5232bac5828d17a6d2e52d4e5091e0877efbd1c7/src/lib.rs#L176). I suspect this is related to your comment in #4, where Mutexes are not able to be tracked properly. I was wondering if you had any tips on how I could solve this?

This is indeed related and how to fix this is an open question. Let me explain a bit more. To better understand the "X depends on Y" mutex graph, we maintain a list of mutexes currently held. This is stored in a thread-local variable:

https://github.com/bertptrs/tracing-mutex/blob/5232bac5828d17a6d2e52d4e5091e0877efbd1c7/src/lib.rs#L87-L93

When a mutex is acquired, its ID is pushed to this stack. When it is released, its ID is removed from the list (with the assumption that it should be near the end). This mechanism works **if and only if** any mutex is released on the same thread it was acquired in. This works fine, until that is the very thing you want to do.

I don't know how to "fix" this yet but I do have some ideas. One is to take a page from the `tracing` crate and add an `instrument` extension trait to a `Future`. This instrumentation could store the list of held mutexes and ensure that it moves between threads as the future itself does. This would make interop between `sync` and `async` mutexes harder, but would at least solve the problem at hand.

I don't think there's a magic bullet, but at least this is where I stand right now. I hope this gives some insight into the problem.

In the meantime I've created #23 to ensure that the thing you tried (and had crash) would result in a compilation failure, rather than a runtime error. I think we can agree that's preferable.

And as a final minor note, could you point your PR at the master branch? The staging branch is where Bors does its job; it's not meant to be used by humans.

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,6 @@
 /target
 Cargo.lock
+
+# IDE folders
+.idea
+.vscode
```
Comment (bertptrs) [inline .gitignore:6]: Please refrain from adding such files to the repository. Adding this for every combination of editor and tooling just results in a mess. I recommend you look into having a global ignores file. You can add one in `~/.config/git/ignore` or wherever your operating system keeps its configuration files.

Comment (mihirsamdarshi) [reply .gitignore:6]: Sorry about that, will do

**Comment by mihirsamdarshi** [comment]
Thanks so much for the guidance, I will look into this and get back to you with a new PR.

**Pull request closed without merging**